utils/src/logging.rs

```diff
@@ -21,3 +29,39 @@ pub fn setup_logging() -> Result<()> {
 
     Ok(())
 }
+
+pub fn format_opt<T: Display>(opt: Option<&T>) -> String {
+    opt.map(ToString::to_string).unwrap_or("None".to_string())
+}
+
+pub fn format_hashes(hashes: &[RpoDigest]) -> String {
+    format!("[{}]", hashes.iter().map(RpoDigest::to_hex).join(", "))
+}
+
+pub fn format_input_notes(notes: &InputNotes<Nullifier>) -> String {
+    format!(
+        "{{ notes: [{}], commitment: {} }}",
+        notes.iter().map(Nullifier::to_hex).join(", "),
+        notes.commitment()
+    )
+}
```

Comment:
A couple of small comments:
- if there are no notes, we should return just "None".
- I think we don't need to log `commitment` field. It is derived from the list of notes.